### PR TITLE
Gemma3 VisionEncoder implementation and checkpoint conversion

### PR DESCRIPTION
**Pull request by hengtaoguo**
# Description

Add Gemma3 Vision Encoder based on original implementation and allow generating matching image embeddings.

1. Building blocks in `Gemma3VisionEncoderLayer` to convert images to image embeddings.
2. Checkpoint conversion in `convert_gemma3_chkpt.py`.
3. Unit test for vision encoder forward pass in `vision_encoder_test.py`. Upload test_image.jpg and golden_data_gemma3_vit.jsonl, which contains `input_image (536,640,3)` and corresponding `soft_embeddings (1,1,256,2560)`.

A follow-up PR needed to change this structure to full comply with MaxText basic layers.

FIXES: [b/410616175](http://b/410616175), [b/411986111](http://b/411986111)

# Tests
Locally tested with matching image_embeddings from the ViT. Newly added unit test suggests the MSE between MaxText image embedding and Gemma3 image embedding is 0.0016501782229170203 smaller than 0.01 threshold.
```
python -m MaxText.decode MaxText/configs/base.yml model_name=gemma3-4b tokenizer_path=assets/tokenizer.gemma3 load_parameters_path=gs://maxtext-model-checkpoints/gemma3-4b/multimodal/2025-04-25-18-06-04/checkpoints/0/items per_device_batch_size=1 run_name=ht_test max_prefill_predict_length=272 max_target_length=300 steps=1 scan_layers=false use_multimodal=True prompt=\'I\ love\ to\ swim\' image_path=\'/home/hengtaoguo/projects/maxtext/MaxText/test_assets/test_image.jpg\' attention=\'dot_product\'
```

```
pytest MaxText/tests/vision_encoder_test.py
```

image

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Review by aireenmei — APPROVED** [review]
Thanks Hengtao!

**Review by gagika — COMMENTED** [review]
I have a few comments,

Could you also run `bash code_style.sh` to make files that you modify pylint complaint?

File: MaxText/convert_gemma3_chkpt.py [hunk 1]
```diff
@@ -76,7 +98,24 @@ def main(raw_args=None) -> None:
...
+              "VisionEmbedder_0": {
+                  "mm_input_projection": params["transformer"]["embedder"]["mm_input_projection"],
+                  "mm_soft_embedding_norm": {
+                      "scale": params["transformer"]["embedder"]["mm_soft_embedding_norm"]["scale"] + 1
+                  },
+              },
+          }
+      },
   }
+  jax_weights = rename_nested_keys(jax_weights, "MlpBlock_0", "MlpBlockViT_0")
```
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply MaxText/convert_gemma3_chkpt.py]: Done by applying layer renaming to only the vision_encoder part and adding more comments.

File: MaxText/layers/gemma3.py [hunk 2]
```diff
@@ -79,8 +79,250 @@ def get_query_pre_attn_scalar(config) -> float:
...
+          (1, shape_product, width),
+          dtype,
+      )
+    elif typ == "sincos2d":
+      return _posemb_sincos_2d(*seqshape, width=width, dtype=dtype)
+    else:
+      raise ValueError(f"Unknown posemb type: {typ}")
 
   @nn.compact
   def __call__(self, inputs, train=False):
```
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply MaxText/layers/gemma3.py]: Done!

File: MaxText/layers/gemma3.py [hunk 3]
```diff
@@ -79,8 +79,250 @@ def get_query_pre_attn_scalar(config) -> float:
...
 class Gemma3VisionEncoderLayer(nn.Module):
   config: Config
+  patch_size: tuple[int] = (14, 14)
+  width: int = 1152
+  mlp_dim: int | None = 4304  # Defaults to 4x input dim
+  depth: int = 27
+  num_heads: int = 16
+  posemb: str = "learn"  # Can also be "sincos2d"
+  dropout: float = 0.0
+  scan: bool = False
```
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply MaxText/layers/gemma3.py]: Done! Using config.scan_layers for the vision_encoder.

File: MaxText/layers/gemma3.py [hunk 4]
```diff
@@ -79,11 +79,252 @@ def get_query_pre_attn_scalar(config) -> float:
...
+    x = x + y
+
+    y = nn.LayerNorm()(x)
+    y = MlpBlockViT(
+        block_id=self.block_id,
+        mlp_dim=self.mlp_dim,
+        dropout=self.dropout,
+        dtype_mm=self.dtype_mm,
+    )(y, deterministic)
+    y = nn.with_logical_constraint(y, ("act_batch", "act_len", "act_emb"))
```
Comment (gobbleturk) [inline MaxText/layers/gemma3.py:161]: can you match these to our rules [here](https://github.com/AI-Hypercomputer/maxtext/blob/2e4af93a82edce7de04ad8a25d09f55313265028/MaxText/configs/base.yml#L276)

"activation_batch"

"activation_length" etc

File: MaxText/layers/gemma3.py [hunk 5]
```diff
@@ -79,11 +79,252 @@ def get_query_pre_attn_scalar(config) -> float:
...
+  width: int = 1152
+  mlp_dim: int | None = 4304  # Defaults to 4x input dim
+  depth: int = 27
+  num_heads: int = 16
+  posemb: str = "learn"  # Can also be "sincos2d"
+  dropout: float = 0.0
+  # or "dots_with_no_batch_dims_saveable" for more speed (memory costly)
+  remat_policy: str = "nothing_saveable"
+  dtype_mm: str = "float32"
+  apply_stop_gradient: bool = True
```
Comment (gobbleturk) [inline MaxText/layers/gemma3.py:301]: can we make this configurable in base.yml

freeze_vision_encoder_params something like that

File: MaxText/layers/gemma3.py [hunk 6]
```diff
@@ -79,11 +79,252 @@ def get_query_pre_attn_scalar(config) -> float:
...
+  patch_size: tuple[int] = (14, 14)
+  width: int = 1152
+  mlp_dim: int | None = 4304  # Defaults to 4x input dim
+  depth: int = 27
+  num_heads: int = 16
+  posemb: str = "learn"  # Can also be "sincos2d"
+  dropout: float = 0.0
+  # or "dots_with_no_batch_dims_saveable" for more speed (memory costly)
+  remat_policy: str = "nothing_saveable"
+  dtype_mm: str = "float32"
```
Comment (gobbleturk) [inline MaxText/layers/gemma3.py:300]: Configurable from base.yml?

new field dtype_mm or rely on the old - I think we want a new field so users can have different dtypes for vision encoder vs the rest

File: MaxText/layers/gemma3.py [hunk 7]
```diff
@@ -92,13 +333,44 @@ def __call__(self, inputs, train=False):
...
+
+    x = nn.Dropout(rate=self.dropout)(x, not train)
+
+    # Transformer encoder to extract image features.
+    x = Encoder(
+        depth=self.depth,
+        mlp_dim=self.mlp_dim,
+        num_heads=self.num_heads,
+        dropout=self.dropout,
+        scan=self.config.scan_layers,
```
Comment (gobbleturk) [inline MaxText/layers/gemma3.py:358]: should scan be configurable from base.yml? Should it always be set to the existing LLM scan?

File: MaxText/layers/gemma3.py [hunk 8]
```diff
@@ -79,11 +79,252 @@ def get_query_pre_attn_scalar(config) -> float:
...
+            mlp_dim=self.mlp_dim,
+            num_heads=self.num_heads,
+            dropout=self.dropout,
+        )
+        x = block_cur(x, deterministic)
+    x: jax.Array = nn.LayerNorm(name="encoder_norm")(x)
+    return x
+
+
+class Einsum(nn.Module):
```
Comment (gobbleturk) [inline MaxText/layers/gemma3.py:220]: Ideally use existing DenseGeneral?

Comment (gobbleturk) [reply MaxText/layers/gemma3.py:220]: This can be a followup PR

File: MaxText/layers/gemma3.py [hunk 9]
```diff
@@ -79,11 +79,252 @@ def get_query_pre_attn_scalar(config) -> float:
...
   config: Config
+  patch_size: tuple[int] = (14, 14)
+  width: int = 1152
+  mlp_dim: int | None = 4304  # Defaults to 4x input dim
+  depth: int = 27
+  num_heads: int = 16
+  posemb: str = "learn"  # Can also be "sincos2d"
+  dropout: float = 0.0
+  # or "dots_with_no_batch_dims_saveable" for more speed (memory costly)
+  remat_policy: str = "nothing_saveable"
```
Comment (gobbleturk) [inline MaxText/layers/gemma3.py:299]: Confirm and add a comment if true: remat policy doesn't matter when params are frozen.

Also use the existing remat policies we have defined

File: MaxText/layers/gemma3.py [hunk 10]
```diff
@@ -92,13 +333,44 @@ def __call__(self, inputs, train=False):
...
+        name="Transformer",
+    )(x, deterministic=deterministic)
+
+    # Gemma3 use a vision exit layer to downsample the soft tokens to a required output length.
+    x = VisionExit(output_length=256)(x)
+    bn, l, c = x.shape
+    x = jnp.reshape(x, [b, n, l, c])
+
+    # VisionEmbedder is a projection layer that projects the image embeddings to align with text embeddings base_emb_dim.
+    x = VisionEmbedder(embed_dim=self.config.base_emb_dim, vision_proj_dim=self.width)(x)
```
Comment (gobbleturk) [inline MaxText/layers/gemma3.py:370]: nit: emb_dim not base_emb_dim

File: MaxText/maxengine.py [hunk 11]
```diff
@@ -325,6 +325,18 @@ def model_apply(_p, _rng):
...
           jnp.ones((1, self.config.max_prefill_predict_length), dtype=jnp.int32),
+          encoder_images=jnp.ones(
+              (
+                  1,
+                  maxtext_utils.NUM_IMAGES_PER_SEQUENCE,
+                  self.config.image_size_for_vit,
+                  self.config.image_size_for_vit,
+                  maxtext_utils.NUM_IMAGE_CHANNELS,
+              ),
+              dtype=jnp.int32,
```
Comment (gobbleturk) [inline MaxText/maxengine.py:336]: is int32 right here? can you include comment?

File: MaxText/tests/vision_encoder_test.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+import jsonlines
+
+from MaxText import pyconfig
+from MaxText import multimodal_utils
+from MaxText.layers import models
+from MaxText.globals import PKG_DIR
+from MaxText import maxengine
+
+
+class VisionEncoderEmbeddingTest(unittest.TestCase):
```
Comment (gobbleturk) [inline MaxText/tests/vision_encoder_test.py:32]: I think this should be an integration XLML test instead?

* This uses real weights pointing a GCS bucket only internal users can run
* it loads real weights which takes a long time

Comment (gobbleturk) [reply MaxText/tests/vision_encoder_test.py:32]: at least move to tests/integration_tests

**Review by gobbleturk — APPROVED** [review]

File: MaxText/layers/gemma3.py (hunk 4 above)
Comment (hengtaoguo) [reply MaxText/layers/gemma3.py]: Done! Thanks for the great catch!

File: MaxText/layers/gemma3.py (hunk 5 above)
Comment (hengtaoguo) [reply MaxText/layers/gemma3.py]: Done! Added a new flag `cfg.freeze_vision_encoder_params=True` as default to stop the gradient in ViT.

File: MaxText/layers/gemma3.py (hunk 6 above)
Comment (hengtaoguo) [reply MaxText/layers/gemma3.py]: Done! Using `cfg.dtype_mm` to control all ViT's basic blocks dtype.

File: MaxText/layers/gemma3.py (hunk 8 above)
Comment (hengtaoguo) [reply MaxText/layers/gemma3.py:220]: This Einsum have learnable weights inside and could be refactor to using DenseGeneral. Let me have a followup PR to refactor the whole Gemma3 ViT using our basic layers.

File: MaxText/layers/gemma3.py (hunk 9 above)
Comment (hengtaoguo) [reply MaxText/layers/gemma3.py]: Added another config.remat_policy_for_vit and set "minimal" as default. This is specifically for the Vision Encoder since it is supposed to be frozen.

File: MaxText/layers/gemma3.py (hunk 10 above)
Comment (hengtaoguo) [reply MaxText/layers/gemma3.py]: Done! Changed all base_emb_dim to emb_dim.

File: MaxText/maxengine.py (hunk 11 above)
Comment (hengtaoguo) [reply MaxText/maxengine.py]: Changed to jnp.float32 to mimic the images after resizing and normalization.

File: MaxText/tests/vision_encoder_test.py (hunk 12 above)
Comment (hengtaoguo) [reply MaxText/tests/vision_encoder_test.py]: Moved to integration_tests, thanks for the great suggestion!

File: MaxText/layers/gemma3.py (hunk 7 above)
Comment (hengtaoguo) [reply MaxText/layers/gemma3.py]: I'm reusing the config.scan_layers to control the ViT. It is set "False" for the unit test and the provided checkpoint is an unscanned version.

**Review by gagika — APPROVED** [review]
thanks!
